### ISSUE TITLE
Tray opens up too late?

### ISSUE THREAD

**Issue by leononame** [Alexays/Waybar#483]
I have an issue where I start KeepassXC and Nextcloud with my sway config. However, waybar startup seems to take much longer than those apps and the tray Icons don't appear unless I quit them and start them manually. 

Is there a solution to this? Something obvious I am missing?

**Comment by alebastr** [comment]
I'm aware of 2 possible solutions:
1. sleep
2. [DBus activation](https://dbus.freedesktop.org/doc/system-activation.txt)

To expand on the second option: you can configure the binary or systemd user service to be started when applications request specific dbus service. [This is](https://github.com/emersion/mako/blob/master/fr.emersion.mako.service.in) how mako defines dbus activation service for notifications, and [Fedora package](https://src.fedoraproject.org/rpms/mako/tree/master) extends that to start mako as a systemd user service.

In this configuration, systemd user service is solving a race condition issue: since waybar now could be started from sway config and from dbus, you need to ensure that only one instance of waybar is running.

I'll try to provide dbus/systemd configs later, when I get to my home pc.

**Comment by leononame** [comment]
Dbus activation sounds great - learned something new. However, is there really a race condition? I can opt out of starting waybar directly from the config file. I can just start the service via sway and KeePassXC and Nextcloud will also start the service by requesting the tray?

**Comment by alebastr** [comment]
DBus service file I'm using:
```ini
# /usr/share/dbus-1/services/org.kde.StatusNotifierWatcher.service
[D-BUS Service]
Name=org.kde.StatusNotifierWatcher
Exec=/usr/bin/waybar
# comment SystemdService to start waybar directly
SystemdService=waybar.service
```
---
While testing this, I found that Qt apps may fail to display icon when waybar is started with dbus service.
This happens because Qt5 [expects](https://github.com/qt/qtbase/blob/e3f5912747af1d46178886a0c5eaaddaf6d1a23b/src/platformsupport/themes/genericunix/dbusmenu/qdbusmenuconnection.cpp#L78) both `org.kde.StatusNotifierWatcher` and `org.kde.StatusNotifierHost` to be ready by the moment of QSystemTrayIcon initialization. Since there's an obvious time lag between the registration of those interfaces, only Watcher is ready when the Qt app checks `IsStatusNotifierHostRegistered`.
As a result, we get  `qt.qpa.menu: StatusNotifierHost is not registered` instead of the icon.

tl;dr: `exec 'sleep 1; nextcloud'` in sway config actually has higher chance to work :disappointed:

**Comment by leononame** [comment]
Alright, thank you. I'll close the issue then.
